nab/detectors/htmcore/htmcore_detector.py

```diff
@@ -0,0 +1,135 @@
+# Copyright (C) 2014, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import math
+
+# htm.core imports
+from htm.bindings.algorithms import SpatialPooler   as SP
+from htm.bindings.algorithms import TemporalMemory  as TM
+from htm.bindings.algorithms import Predictor
+
+from htm.algorithms.anomaly_likelihood import AnomalyLikelihood
+from nab.detectors.base import AnomalyDetector
+
+# Fraction outside of the range of values seen so far that will be considered
+# a spatial anomaly regardless of the anomaly likelihood calculation. This
+# accounts for the human labelling bias for spatial values larger than what
+# has been seen so far.
+SPATIAL_TOLERANCE = 0.05
+
+
+
+class HtmcoreDetector(AnomalyDetector):
+  """
+  This detector uses an HTM based anomaly detection technique.
+  """
+
+  def __init__(self, *args, **kwargs):
+
+    super(HtmcoreDetector, self).__init__(*args, **kwargs)
+
+    self.anomalyLikelihood = None
+    # Keep track of value range for spatial anomaly detection
+    self.minVal = None
+    self.maxVal = None
+
+    # Set this to False if you want to get results based on raw scores
+    # without using AnomalyLikelihood. This will give worse results, but
+    # useful for checking the efficacy of AnomalyLikelihood. You will need
+    # to re-optimize the thresholds when running with this setting.
+    self.useLikelihood = True
+
+
+  def getAdditionalHeaders(self):
+    """Returns a list of strings."""
+    return ["raw_score"] #TODO add "prediction"
+
+
+  def handleRecord(self, inputData):
+    """Returns a tuple (anomalyScore, rawScore).
+
+    Internally to NuPIC "anomalyScore" corresponds to "likelihood_score"
+    and "rawScore" corresponds to "anomaly_score". Sorry about that.
+    """
+    # Send it to Numenta detector and get back the results
+    result = [] #FIXME self.model.run(inputData)
+
+    # Get the value
+    value = inputData["value"]
+
+    # Retrieve the anomaly score and write it to a file
+    rawScore = 0.5 #FIXME result.inferences["anomalyScore"]
+
+    # Update min/max values and check if there is a spatial anomaly
+    spatialAnomaly = False #TODO make this computed in SP (and later improve)
```

Comment:
- I like the idea of SpatialAnomaly being moved to SP itself https://github.com/htm-community/htm.core/issues/550
- I don't 100% like the current implementation of spatial anomalies, but that can be improved later.

Reply:
done in https://github.com/htm-community/htm.core/pull/587